clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/id/DcosSpinnakerAppId.groovy

```diff
@@ -0,0 +1,153 @@
+package com.netflix.spinnaker.clouddriver.dcos.deploy.util.id
+
+import static com.google.common.base.Preconditions.checkArgument
+import static com.google.common.base.Strings.nullToEmpty
+
+import com.netflix.frigga.Names
+
+/**
+ * Represents a hierarchical Spinnaker specific application identifier for DCOS.
+ * Structure - /account/region/app-stack-detail-sequence
+ */
+class DcosSpinnakerAppId {
+    public final static def SAFE_REGION_SEPARATOR = "_"
+
+    private final def marathonPath
+
+    public DcosSpinnakerAppId(final String marathonAppId) {
```

Comment:
The only reason he lists that would actually apply here is the first advantage, and allowing for naming seems only really useful when the object being created is special (in other words, the object you are transforming to is important rather than the objects you are transforming from) or if there is collision with parameters (which I actually ran into with LBId so I will go with static factory methods for both instead of public constructors).

As for optionals, we shouldn't be relying on outside validation for constructing an object. An object creator (our static factory methods) should be ensuring they are valid, not whoever is calling this.